torch/csrc/jit/passes/peephole.cpp

```diff
@@ -165,6 +165,51 @@ void PeepholeOptimizeImpl(Block* block, bool addmm_fusion_enabled) {
           u.user->replaceInput(0, node->inputs().at(0));
         }
       }
+    } else if (node->kind() == prim::If) {
+      IfView n(node);
+      // this handles redundant short circuits like "x and True" or "x or False"
+      for (size_t i = 0; i < n.outputs().size(); ++i) {
+        if (n.outputs().at(i)->type() != BoolType::get()) {
+          continue;
+        }
+        bool true_val =
+            constant_as<bool>(n.thenOutputs().at(i)).value_or(false);
+        bool false_val =
+            constant_as<bool>(n.elseOutputs().at(i)).value_or(true);
+        // if an if node's output equals its condition replace output with
+        // condition
+        if (true_val && !false_val) {
+          n.outputs().at(i)->replaceAllUsesWith(n.cond());
+        }
+      }
+    } else if (
+        node->kind() == aten::__is__ || node->kind() == aten::__isnot__) {
+      // if we are comparing a None value with a value that can't be None
+      // replace the output with true if node is __isnot__ or false if node is
+      // __is__
+      AT_ASSERT(node->inputs().size() == 2);
+      for (size_t check_none_index : {0, 1}) {
```

Comment:
Why do you only check for must be none + must not be none and not for must be none and must be none?

Reply:
The two must be none cases is already handled by constant propagation. I'll add a comment

Reply:
Ah, yes, but that is only for constants. :slightly_smiling_face:  Should we decide to specialize graphs based on `None` in `Optional[Tensor]` arguments, this would not be covered by const prop. Until then, I can see there is no need.

Reply:
Up to this point if a node mustBeNone it's a constant. I still don't super understand the AutoGradZero case. If we can treat it as a constant I think we should modify constant propagation so that is the general behavior instead of just here @wanchaol